src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java

```diff
@@ -273,11 +275,12 @@ public void evictAllGraphsFromCache() {
      *
      * @param indexPathUrl path for serialized hnsw graph
      * @param indexName index name
+     * @param efSearch HNSW parameter that represents "the size of the dynamic list for the nearest neighbors"
      * @return KNNIndex holding the heap pointer of the loaded graph
      * @throws Exception Exception could occur when registering the index path
      * to Resource watcher or if the JNI call throws
      */
-    public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName) throws Exception {
+    public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName, int efSearch) throws Exception {
```

Comment:
This method would be called only once when there is cache miss. The ```efsearch``` value passed is stored in the 
[IndexWrapper](https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/jni/src/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp#L49) as part of [init](https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/jni/src/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp#L204) call. So ef_search passed in subsequent queries would not be reflected. 

Ideal way is to plumb the ef_search through [queryIndex](https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/knn/index/v1736/KNNIndex.java#L71) function from [KNNWeight](https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java#L101). As part of this function in the [jni](https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/jni/src/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp#L147), we could set the  ```ef_search``` using```setQueryTimeParams``` if ```ef_search``` is not null otherwise just use the existing value. 

We need to be careful while calling ```setQueryTimeParams``` because at the same time it is possible some query is already referencing this object and could cause segmentation faults. We need to kind of ensure only one thread is setting/getting ```ef_search``` .

Reply:
> Ideal way is to plumb the ef_search through queryIndex function from KNNWeight. As part of this function in the jni, we could set the ef_search using setQueryTimeParams if ef_search is not null otherwise just use the existing value.

Thanks for pointing that out. I need to renew license on CLion for getting this part done.

> We need to be careful while calling setQueryTimeParams because at the same time it is possible some query is already referencing this object and could cause segmentation faults. We need to kind of ensure only one thread is setting/getting ef_search .

Thanks for reminder. My first thought is to use writeLock in queryIndex if ef_search is not null. Might need to put some thought into this. Suggestion is welcome.

Reply:
I suspect that supporting a dynamic efSearch parameter would either require:
1. Synchronizing search queries from multiple threads onto a shared index (which would likely hurt query performance significantly)
2. Require changes to the KNN library dependency to make the efSearch parameter a local parameter of the KNN query structure rather than a member of the index. It's currently a class member of the [HNSW index](https://github.com/nmslib/nmslib/blob/master/similarity_search/include/method/hnsw.h#L519), which is why the opendistro plugin is setting it once on index load rather than supporting per query, since mutations within queries would affect unrelated queries (see [here](https://github.com/nmslib/nmslib/blob/master/similarity_search/src/method/hnsw.cc#L1205) for one example usage of the member variable).

Reply:
Agree. Synchronizing could hurt performance because of locking. At this point to me, Approach 2 mentioned by @jschmitz28  seems to be best way to achieve this.

Other solution i could think of is, make ef_search setting dynamic setting(currently its static, once define cannot be changed). On the change of this setting, we could evict the graphs from memory and load freshly with the new ef_search.

Reply:
@vamshin I have two more points to clarify regarding your solution, first UX and second implementation:

(1) dynamical setting still needs to be updated through index settings API: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html
so I think UX is going to be different from #116 . Is that OK? Otherwise I will return to approach 2 by @jschmitz28 .

(2) On trying to implemnt your solution by defining ef_search to be index scope and dynamic. As you mentioned, we also need to add custom setting update consumer for that ef_search. Now the issue is that IndexModule has not yet been passed into KNNSettings so we do not have way to add the consumer for index scope settings. 

From my observation, we can implement

https://github.com/elastic/elasticsearch/blob/853dd1b873d02d04231c32722d4784c9da1068dc/server/src/main/java/org/elasticsearch/plugins/Plugin.java#L131

in KNNPlugin, then we add an accept index module API in KNNSetting. Thought?

Reply:
@chenqi0805 

(1) Seems to be perfectly ok to change that. 
Regarding the approach mentioned by @jschmitz28  we could create an issue in the nmslib and ask their opinion about this change? Not sure if they had to do this way for some optimizations.

(2) Makes sense.

Reply:
https://github.com/nmslib/nmslib/issues/442